pkg/theiactl/commands/check.go

```diff
@@ -0,0 +1,207 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package commands
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"antrea.io/theia/pkg/theiactl/portforwarder"
+	sparkv1 "antrea.io/theia/third_party/sparkoperator/v1beta2"
+)
+
+// checkCmd represents the check command
+var checkCmd = &cobra.Command{
+	Use:   "check",
+	Short: "Check the status of a policy recommendation Spark job",
+	Long: `Check the current status of a policy recommendation Spark job by ID.
+It will return the status of this Spark application like SUBMITTED, RUNNING, COMPLETED, or FAILED.`,
+	Example: `
+Check the current status of job with ID e998433e-accb-4888-9fc8-06563f073e86
+$ theiactl policyreco check --id e998433e-accb-4888-9fc8-06563f073e86
+`,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		recoID, err := cmd.Flags().GetString("id")
+		if err != nil {
+			return err
+		}
+		_, err = uuid.Parse(recoID)
+		if err != nil {
+			return fmt.Errorf("failed to decode input id %s into a UUID, err: %v", recoID, err)
+		}
+		kubeconfig, err := cmd.Flags().GetString("kubeconfig")
+		if err != nil {
+			return err
+		}
+		clientset, err := CreateK8sClient(kubeconfig)
+		if err != nil {
+			return fmt.Errorf("couldn't create k8s client using given kubeconfig, %v", err)
+		}
+		useClusterIP, err := cmd.Flags().GetBool("use-cluster-ip")
+		if err != nil {
+			return err
+		}
+
+		err = PolicyRecoPreCheck(clientset)
+		if err != nil {
+			return err
+		}
+
+		sparkApplication := &sparkv1.SparkApplication{}
+		err = clientset.CoreV1().RESTClient().
+			Get().
+			AbsPath("/apis/sparkoperator.k8s.io/v1beta2").
+			Namespace(flowVisibilityNS).
+			Resource("sparkapplications").
+			Name("policy-reco-" + recoID).
+			Do(context.TODO()).
+			Into(sparkApplication)
+		if err != nil {
+			return err
+		}
+		state := strings.TrimSpace(string(sparkApplication.Status.AppState.State))
+		if state == "RUNNING" {
+			var endpoint string
+			service := fmt.Sprintf("policy-reco-%s-ui-svc", recoID)
+			if useClusterIP {
+				serviceIP, servicePort, err := GetServiceAddr(clientset, service)
+				if err != nil {
+					return fmt.Errorf("error when getting Spark Monitor Service address: %v", err)
+				}
+				endpoint = fmt.Sprintf("tcp://%s:%d", serviceIP, servicePort)
+			} else {
+				servicePort := 4040
+				listenAddress := "localhost"
```

Comment:
I think we wanted this to be k8s API IP so that theiactl could be run externally? (same in other places)

Reply:
I thought we were forwarding the service port to the machine which runs theiactl? What does K8s API IP refers to?

Reply:
AFAIU we need to expose the port on k8s server API (Host in k8s config)

Reply:
I don't think kubectl port forward will open port on the K8S API server, will it?
That seems the wrong thing to do from a security perspective.

I seem to recall it's

localhost:localport <---> K8SAPISERVER <---> Target Pod:Target port

Reply:
Yes, sorry, you're both right, API server acts as a proxy here.

Reply:
Thanks for the explanation!